File: Java/014.java
import java.util.HashMap;
import java.util.Map;

public class Problem014 {

    private static final Map<Long, Integer> lengths = new HashMap<>();

    public static int collatzLength(long n) {
        if (lengths.containsKey(n)) {
            return lengths.get(n);
        }
        if (n % 2 == 0) {
            lengths.put(n, 1 + collatzLength(n / 2));
        } else {
            lengths.put(n, 1 + collatzLength(3 * n + 1));
        }
        return lengths.get(n);
    }

    public static void main(String[] args) {
        lengths.put(1L, 1);
        long numWithLongestLength = 0;
        int longestLength = 0;
        for (long n = 1; n < 1_000_000; n++) {
            int currentLength = collatzLength(n);
            if (currentLength > longestLength) {
                numWithLongestLength = n;
                longestLength = currentLength;
            }
        }
        System.out.println(numWithLongestLength);
    }
}

File: Rust/014.rs
use std::collections::HashMap;

fn collatz_length(n: u64, lengths: &mut HashMap<u64, u32>) -> u32 {
    if let Some(&len) = lengths.get(&n) {
        return len;
    }

    let len = if n % 2 == 0 {
        1 + collatz_length(n / 2, lengths)
    } else {
        1 + collatz_length(3 * n + 1, lengths)
    };

    lengths.insert(n, len);
    len
}

fn main() {
    let mut lengths: HashMap<u64, u32> = HashMap::new();
    lengths.insert(1, 1);

    let mut num_with_longest_length: u64 = 0;
    let mut longest_length: u32 = 0;

    for n in 1..1_000_000 {
        let current_length = collatz_length(n, &mut lengths);
        if current_length > longest_length {
            longest_length = current_length;
            num_with_longest_length = n;
        }
    }

    println!("{}", num_with_longest_length);
}
